### PR TITLE
User + Role

### PR DESCRIPTION
**Pull request by Jakarai**
## What does this PR do?
Added user and Roles system

## How can others know that your changes do what you say they do? Can you please share some testing steps?

After running the main class, you can then go to 'Localhost:9001/api/prankers', then you should see a login screen generated by Spring. enter 'user' for the username and use the spring generated password, as the password. you should than see a JSON array of users.

Screen Shot 2022-03-18 at 3 22 19 PM

## Checklist before merging
- [ ] Someone has given this PR an approval and thorough review
- [ ] I have validated that I have not broken something else in
Screen Shot 2022-03-18 at 3 20 15 PM
 the code

### CODE REVIEW

**Review by chumomega — CHANGES_REQUESTED** [review]
bro you ain't notice that this pr is adding 2million lines of code lol?
the pranker frontend stuff doesn't belong in this repo. and even if it did, we don't need the node_modules since that can be geneerated locally using npm install

**Comment by Jakarai** [comment]
Nah this is madness, I had no clue All that code was being added, not sure how it even got there.

**Comment by Jakarai** [comment]
Im guessing my dev branch still had the old code with the frontend still attached instead of separate repos.

File: src/main/resources/application.properties [hunk 1]
```diff
@@ -1,11 +1,13 @@
-server.port=8990
+server.port=9001
```
Comment (chumomega) [inline src/main/resources/application.properties:1]: why are you changing the port?

File: src/main/java/com/v2/pranker/PrankerApplication.java [hunk 2]
```diff
@@ -21,4 +26,26 @@
 	public static void main(String[] args) {
 		SpringApplication.run(PrankerApplication.class, args);
 	}
+
+	@Bean
+	CommandLineRunner run(PrankerService prankerService) {
+		return args -> {
+			prankerService.saveRole(new Role(null, "Role_User"));
```
Comment (chumomega) [inline src/main/java/com/v2/pranker/PrankerApplication.java:33]: these run sql commands in the background right? could we try to remove this and just add this mock data locally (either some sql script or manually)?

File: src/main/java/com/v2/pranker/controller/PrankerController.java [hunk 3]
```diff
@@ -0,0 +1,49 @@
...
+import java.net.URI;
+import java.util.List;
+
+@RequiredArgsConstructor
+@RequestMapping("/api")
+@RestController
+public class PrankerController {
+    private final PrankerService prankerService;
+
+    @GetMapping("/prankers")
```
Comment (chumomega) [inline src/main/java/com/v2/pranker/controller/PrankerController.java:22]: why would we need to get all prankers?

File: src/main/java/com/v2/pranker/controller/PrankerController.java [hunk 4]
```diff
@@ -0,0 +1,49 @@
...
+    public ResponseEntity<List<Pranker>>getPrankers() {
+        return ResponseEntity.ok().body(prankerService.getPrankers());
+    }
+    @PostMapping("/pranker/save")
+    public ResponseEntity<Pranker>savePranker(@RequestBody Pranker pranker){
+        URI uri = URI.create(ServletUriComponentsBuilder.fromCurrentContextPath().path("/api/user/save").toUriString());
+        return ResponseEntity.created(uri).body(prankerService.savePranker(pranker));
+    }
+    @PostMapping("/role/save")
+    public ResponseEntity<Role>saveRole(@RequestBody Role role) {
```
Comment (chumomega) [inline src/main/java/com/v2/pranker/controller/PrankerController.java:32]: why do we need to publicly expose the creating of a new role via?

File: src/main/java/com/v2/pranker/controller/PrankerController.java [hunk 5]
```diff
@@ -0,0 +1,49 @@
...
+        URI uri = URI.create(ServletUriComponentsBuilder.fromCurrentContextPath().path("/api/role/save").toUriString());
+        return ResponseEntity.created(uri).body(prankerService.saveRole(role));
+    }
+    @PostMapping("/role/addtouser")
+    public ResponseEntity<?>addRoleToPranker(@RequestBody RoleToPrankerForm form) {
```
Comment (chumomega) [inline src/main/java/com/v2/pranker/controller/PrankerController.java:37]: why do we need to publicly expose the adding of a role to a user?

**Review by Jakarai — COMMENTED** [review]
no reason, I changed it during running multiple servers, I changed back

File: src/main/java/com/v2/pranker/repos/PrankerRepo.java [hunk 6]
```diff
@@ -0,0 +1,11 @@
+package com.v2.pranker.repos;
+
+import com.v2.pranker.model.Pranker;
+
+import org.springframework.data.jpa.repository.JpaRepository;
+
+public interface PrankerRepo extends JpaRepository<Pranker, Long> {
```
Comment (chumomega) [inline src/main/java/com/v2/pranker/repos/PrankerRepo.java:7]: why do we need an interface, can't this just contain all the methods you implemented in the `...service` file

File: src/main/java/com/v2/pranker/services/PrankerServiceImpl.java [hunk 7]
```diff
@@ -0,0 +1,47 @@
...
+        return prankerRepo.save(pranker);
+    }
+
+    @Override
+    public Role saveRole(Role role) {
+        return roleRepo.save(role);
+    }
+
+    @Override
+    public void addRoleToPranker(String prankername, String roleName) {
```
Comment (chumomega) [inline src/main/java/com/v2/pranker/services/PrankerServiceImpl.java:32]: [nit] your variable names aren't consistent

**Review by chumomega — CHANGES_REQUESTED** [review]
this pr is huge, adds a lot of fundamental changes to how we access the database, and adds a major change in how the project is structured. i don't necessarily think it's wrong or right, but i do think it will make contributing to this project much more complex. 
Path forward: I think we need to come to a consensus on how we want to structure the project, access the database, before moving forward. If we're going to move from pure JDBC to JPA, then we need to do a whole refactor so there is consistency.

File: src/main/java/com/v2/pranker/controller/PrankerController.java (hunk 4 above)
Comment (Jakarai) [reply src/main/java/com/v2/pranker/controller/PrankerController.java:32]: Visibility for Spring beans

File: src/main/java/com/v2/pranker/services/PrankerServiceImpl.java (hunk 7 above)
Comment (Jakarai) [reply src/main/java/com/v2/pranker/services/PrankerServiceImpl.java:32]: These are also already refactored, I wanted to PR in by working parts as opposed to one big PR

**Comment by Jakarai** [comment]
Btw JPA is built on top of JDBC its handling all the ORM for me.

File: src/main/java/com/v2/pranker/controller/PrankerController.java (hunk 4 above)
Comment (chumomega) [reply src/main/java/com/v2/pranker/controller/PrankerController.java:50]: what do you mean by that?

**Comment by chumomega** [comment]
could we convert to using native jdbc for now? i know jpa has certain benefits, but I'm thinking about ease of development and neither myself nor karen are super familiar with jpa. also(and more importantly imo), we already have the entire codebase built on native jdbc. adding this without a large scale refactor would fracture the codebase and make it more complex.

**Review by chumomega — APPROVED** [review]
approved, but i still feel icky about how huge this pr is and how much is being added without fully grasping what each piece is doing

**Pull request merged**